functionary/core/auth/__init__.py

```diff
@@ -0,0 +1,84 @@
+from enum import Enum
+
+
+class Permissions(Enum):
+    """Enum containing permissions, which consist of one for each of the CRUD
+    operations per model. Examples:
+
+        PACKAGE_CREATE = "package:create"
+        PACKAGE_READ = "package:read"
+        PACKAGE_UPDATE = "package:update"
+        PACKAGE_DELETE = "package:delete"
+
+    These are intended to be used with the User has_perm() method and can be provided
+    as either the enum or its value:
+
+        # both of these are valid
+        user.has_perm(Permissions.FUNCTION_READ, environment)
+        user.has_perm(Permissions.FUNCTION_READ.value, environment)
+    """
+
+    ENVIRONMENT_CREATE = "environment:create"
+    ENVIRONMENT_READ = "environment:read"
+    ENVIRONMENT_UPDATE = "environment:update"
+    ENVIRONMENT_DELETE = "environment:delete"
+
+    FUNCTION_CREATE = "function:create"
+    FUNCTION_READ = "function:read"
+    FUNCTION_UPDATE = "function:update"
+    FUNCTION_DELETE = "function:delete"
```

Comment:
What are the constraints for a user to execute a function? If they can see it and have an environment, they can execute it?

Reply:
Right now I'm still thinking that at some point soon we'll have a `Task` model, which would have its own set of CRUD permissions here.  So executing a function would be governed by the `task:create` permission.